src/main/org/apache/tools/ant/taskdefs/optional/junitlauncher/JUnitLauncherTask.java

```diff
@@ -0,0 +1,508 @@
+                    errorMessage = "Some test(s) have failure(s)";
+                }
+                throw new BuildException(errorMessage);
+            }
+        }
+    }
+
+    private ClassLoader createClassLoaderForTestExecution() {
+        if (this.classPath == null) {
+            return this.getClass().getClassLoader();
+        }
+        return new AntClassLoader(this.getClass().getClassLoader(), getProject(), this.classPath, true);
+    }
+
+    private Optional<SwitchedStreamHandle> trySwitchSysOut(final TestRequest testRequest) {
+        if (!testRequest.interestedInSysOut()) {
+            return Optional.empty();
+        }
+        final PipedOutputStream pipedOutputStream = new PipedOutputStream();
+        final PipedInputStream pipedInputStream;
+        try {
+            pipedInputStream = new PipedInputStream(pipedOutputStream);
+        } catch (IOException ioe) {
+            // log and return
+            return Optional.empty();
+        }
+
+        final PrintStream printStream = new PrintStream(pipedOutputStream, true);
+        System.setOut(new PrintStream(printStream));
+
+        final SysOutErrStreamReader streamer = new SysOutErrStreamReader(this, pipedInputStream,
+                StreamType.SYS_OUT, testRequest.getSysOutInterests());
+        final Thread sysOutStreamer = new Thread(streamer);
+        sysOutStreamer.setDaemon(true);
+        sysOutStreamer.setName("junitlauncher-sysout-stream-reader");
+        sysOutStreamer.setUncaughtExceptionHandler((t, e) -> this.log("Failed in sysout streaming", e, Project.MSG_INFO));
+        sysOutStreamer.start();
+        return Optional.of(new SwitchedStreamHandle(pipedOutputStream, streamer));
+    }
+
+    private Optional<SwitchedStreamHandle> trySwitchSysErr(final TestRequest testRequest) {
+        if (!testRequest.interestedInSysErr()) {
+            return Optional.empty();
+        }
+        final PipedOutputStream pipedOutputStream = new PipedOutputStream();
+        final PipedInputStream pipedInputStream;
+        try {
+            pipedInputStream = new PipedInputStream(pipedOutputStream);
+        } catch (IOException ioe) {
+            // log and return
+            return Optional.empty();
+        }
+
+        final PrintStream printStream = new PrintStream(pipedOutputStream, true);
+        System.setErr(new PrintStream(printStream));
+
+        final SysOutErrStreamReader streamer = new SysOutErrStreamReader(this, pipedInputStream,
+                StreamType.SYS_ERR, testRequest.getSysErrInterests());
+        final Thread sysErrStreamer = new Thread(streamer);
+        sysErrStreamer.setDaemon(true);
+        sysErrStreamer.setName("junitlauncher-syserr-stream-reader");
+        sysErrStreamer.setUncaughtExceptionHandler((t, e) -> this.log("Failed in syserr streaming", e, Project.MSG_INFO));
+        sysErrStreamer.start();
+        return Optional.of(new SwitchedStreamHandle(pipedOutputStream, streamer));
+    }
+
```

Comment:
you could probably DRY up the two methods a little by extracting the common logic.

Reply:
Yes makes sense, I'll clean up this part a bit.